pydatalab/src/pydatalab/routes/v0_1/items.py

```diff
@@ -1048,6 +1067,418 @@ def get_item_data(
+    except ValidationError as exc:
+        LOGGER.error(
+            "Version counter validation failed for refcode %s: %s",
+            refcode,
+            str(exc),
+        )
+        # Fallback: return raw counter value to prevent blocking saves
+        # This should only happen if the document is corrupted
+        return result["counter"]
+
+
+@ITEMS.route("/items/<refcode>/versions/", methods=["GET"])
+def list_versions(refcode):
+    """List all saved versions for an item, with metadata."""
+    if len(refcode.split(":")) != 2:
+        refcode = f"{CONFIG.IDENTIFIER_PREFIX}:{refcode}"
+
+    versions = list(
+        flask_mongo.db.item_versions.find(
+            {"refcode": refcode},
+            {
+                "_id": 1,
+                "timestamp": 1,
+                "user_id": 1,
+                "datalab_version": 1,
+                "version": 1,
+                "action": 1,
+                "restored_from_version": 1,
+                "data.version": 1,
+            },
+        ).sort("version", -1)
+    )
+    for v in versions:
+        v["_id"] = str(v["_id"])
+    return jsonify({"status": "success", "versions": versions}), 200
+
+
+@ITEMS.route("/items/<refcode>/versions/<version_id>/", methods=["GET"])
+def get_version(refcode, version_id):
+    """Get the full snapshot of an item at a specific version."""
+    if len(refcode.split(":")) != 2:
+        refcode = f"{CONFIG.IDENTIFIER_PREFIX}:{refcode}"
+
+    try:
+        version_object_id = ObjectId(version_id)
+    except (InvalidId, TypeError):
+        return jsonify({"status": "error", "message": f"Invalid version_id: {version_id}"}), 400
+
+    version = flask_mongo.db.item_versions.find_one({"_id": version_object_id, "refcode": refcode})
+    if not version:
+        return jsonify({"status": "error", "message": "Version not found"}), 404
+    version["_id"] = str(version["_id"])
+    return jsonify({"status": "success", "version": version}), 200
+
+
+@ITEMS.route("/items/<refcode>/compare-versions/", methods=["GET"])
+def compare_versions(refcode):
+    """Compare two versions of an item and return their differences, including version numbers."""
+    if len(refcode.split(":")) != 2:
+        refcode = f"{CONFIG.IDENTIFIER_PREFIX}:{refcode}"
+
+    # Validate query parameters using Pydantic model
+    try:
+        query_params = CompareVersionsQuery(
+            v1=request.args.get("v1", ""), v2=request.args.get("v2", "")
+        )
+    except ValidationError as exc:
+        return jsonify(
+            {"status": "error", "message": "Invalid query parameters", "errors": exc.errors()}
+        ), 400
+
+    try:
+        v1_object_id = ObjectId(query_params.v1)
+        v2_object_id = ObjectId(query_params.v2)
+    except (InvalidId, TypeError) as e:
+        return jsonify({"status": "error", "message": f"Invalid version ID format: {str(e)}"}), 400
+
+    v1 = flask_mongo.db.item_versions.find_one({"_id": v1_object_id, "refcode": refcode})
```

Comment:
Likewise in this route, need to check that the user has read permissions on the sample first